реализованы абстракция персонажей и классы воина и лучника

licosty opened the pull request.

licosty committed:
реализованы абстракция персонажей и классы воина и лучника

zaqqaz226 commented on src/main/java/school/faang/bjs2_79166/Main.java:
```diff
@@ -0,0 +1,14 @@
+        Character warrior = new Warrior("John");
+        Character archer = new Archer("Jane");
+
+        while (true) {
```
Мы тут попадем в вечный цикл. Может лучше проверять пока один из персонажей не останется без жизней?)

zaqqaz226 commented on src/main/java/school/faang/bjs2_79166/Archer.java:
```diff
@@ -0,0 +1,19 @@
+            health = 0;
+            throw new RuntimeException(opponent.name + " is killed");
+        } else {
+            System.out.println(String.format("%s наносит урон %s равный %d", this.name, opponent.name, this.agility));
```
System.out.println конечно хорошо, но если ты на класс накинешь аннотацию @Slf4j, то в этом месте сможешь использовать не вывод в консоль, а логирование, что гораздо лучше. Логирование делается после накидывания аннотации и использования log.info("Your message");

zaqqaz226 commented on src/main/java/school/faang/bjs2_79166/Archer.java:
```diff
@@ -0,0 +1,19 @@
+    public void attack(Character opponent) {
+        if (isKilled(opponent.health, agility)) {
+            health = 0;
+            throw new RuntimeException(opponent.name + " is killed");
```
Если переделать цикл в main, то тут уже не понадобиться выкидывать исключения для остановки работы.

zaqqaz226 commented on src/main/java/school/faang/bjs2_79166/Warrior.java:
```diff
@@ -0,0 +1,19 @@
+package school.faang.bjs2_79166;
+
+
+public class Warrior extends Character {
```
Та же история, что и в другом классе.

licosty committed:
- System.out.println заменен на логирование
- переделан цикл в main
- attack больше не бросает исключение, если health <= 0

licosty committed:
- скорректировано уменьшение здоровья в attack
- скорректирован main

zaqqaz226 commented on src/main/java/school/faang/bjs2_79166/Main.java:
```diff
@@ -0,0 +1,21 @@
+        Character warrior = new Warrior("John");
+        Character archer = new Archer("Jane");
+
+        while (true) {
```
Так как цикл while будет постоянно выполняться и проверять условие, которое у него внутри, можно просто прописать условия из if`ов прямо в нем. Когда условие станет false, то while закончит работу

licosty replied:
@zaqqaz226 Имеется ввиду if (archer.health > 0 && warrior.health > 0) ?
Но тогда в последней итерации герои убивают друг друга.

zaqqaz226 replied:
Ты имеешь ввиду, чтобы мертвый герой уже не атаковал? Да, согласен

zaqqaz226 commented on src/main/java/school/faang/bjs2_79166/Warrior.java:
```diff
@@ -0,0 +1,21 @@
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
```
Класс ) Смотрится намного лучше )
